external/llvm-project/mlir/lib/Transforms/Utils/OutlinerUtils.cpp

```diff
@@ -173,337 +209,380 @@ static bool opsMatch(Operation *lhs, Operation *rhs, OutliningCandidate &one,
+    auto candidate = std::get<0>(candidatePair);
+    if (std::get<1>(candidatePair)) {
+      // Build outlined func.
+      // And update the candidate with the cloned ops for comparison.
+      candidate->setFunction(build(candidates.size() - 1, attrName));
+    }
+    return candidate->getFunction();
+  }
 
-    // Insert outlined function before current function.
-    OpBuilder::InsertionGuard g(b);
-    b.setInsertionPoint(anchorOp->getParentOfType<func::FuncOp>());
+  // Build the outlined function.
+  func::FuncOp build(uint32_t idx, StringRef attrName) const {
+    func::FuncOp anchorFunc = _anchorOp->getParentOfType<func::FuncOp>();
+    ValueRange inputs(_inputs.getArrayRef());
+    OpBuilder b(anchorFunc);
+    Location loc = _anchorOp->getLoc();
+    MLIRContext *ctx = _anchorOp->getContext();
+
+    auto partFnName =
+        anchorFunc.getSymName().str() + "__part_" + std::to_string(idx);
 
     // Make FuncOp from anchorOp's operand types and trailingOp's result type.
-    MLIRContext *ctx = anchorOp->getContext();
-    ValueRange results(returnVals);
+    ValueRange results(_results.getArrayRef());
     FunctionType type =
-        FunctionType::get(ctx, values.getTypes(), results.getTypes());
+        FunctionType::get(ctx, inputs.getTypes(), results.getTypes());
     SmallVector<NamedAttribute, 1> kernelAttrs{
         b.getNamedAttr(attrName, b.getUnitAttr()),
     };
-    outlinedFunc = b.create<func::FuncOp>(
+    func::FuncOp outlinedFunc = b.create<func::FuncOp>(
         loc, partFnName, type, ArrayRef<NamedAttribute>(kernelAttrs));
     outlinedFunc->setAttr("sym_visibility", StringAttr::get(ctx, "private"));
-    newCandidate.function = outlinedFunc;
 
     // Add access modes for parameters: read-only, write-only, read-write
     // All MemRef params are marked as 'read-write'
     // Non-MemRef inputs are added as 'read-only'
-    auto readAccessAttr =
+    auto readAttr =
         b.getNamedAttr(func::FuncOp::getReadAccessAttrName(), b.getUnitAttr());
-    auto writeAccessAttr =
+    auto writeAttr =
         b.getNamedAttr(func::FuncOp::getWriteAccessAttrName(), b.getUnitAttr());
-    for (auto pair : llvm::enumerate(values)) {
-      auto vtype = pair.value().getType();
-      if (vtype.isa<VectorType, RankedTensorType, UnrankedTensorType>())
-        outlinedFunc.setArgAttrs(pair.index(),
-                                 b.getDictionaryAttr({readAccessAttr}));
-      else if (vtype.isa<MemRefType>())
-        outlinedFunc.setArgAttrs(
-            pair.index(),
-            b.getDictionaryAttr({readAccessAttr, writeAccessAttr}));
+    auto getAccessAttrs = [&](Type t,
+                              bool inputs) -> std::optional<DictionaryAttr> {
+      if (t.isa<VectorType, RankedTensorType, UnrankedTensorType>())
+        return b.getDictionaryAttr({inputs ? readAttr : writeAttr});
+      if (t.isa<MemRefType>())
+        return b.getDictionaryAttr({readAttr, writeAttr});
```

Comment:
Is this overly conservative, or am I misremembering when this code runs?

Reply:
Actually I am now disallowing MemRefTypes so it could be removed.